src/java.base/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java

```diff
@@ -147,6 +147,14 @@ public void cleanup() {
         return segment;
     }
 
+    private static long allocateMemoryWrapper(long size) {
+        try {
+            return UNSAFE.allocateMemory(size);
+        } catch (IllegalArgumentException ex) {
+            throw new OutOfMemoryError();
```

Comment:
`OutOfMemoryError` should be updated to have the `Throwable`‑accepting constructor overloads, so that this can include the cause:
```suggestion
            throw new OutOfMemoryError(ex);
```

See also: https://github.com/openjdk/panama-foreign/pull/855#discussion_r1285058300

Reply:
I think enhancing `OutOfMemoryError` is a discussion to be had separately. Not as a part of this JEP.